### PR TITLE
First pass at invite-only migration.

### PR DESCRIPTION
**Pull request by dessalines**

### CODE REVIEW

File: migrations/2021-11-23-153753_add_invite_only_columns/up.sql
Comment (dessalines) [inline migrations/2021-11-23-153753_add_invite_only_columns/up.sql:19]: @Nutomic take a look at these, I changed some of the column names to make more sense. If they look good I can continue on with it.

Comment (Nutomic) [reply migrations/2021-11-23-153753_add_invite_only_columns/up.sql:19]: `published` is the creation time? That one is a bit unclear, others are fine. Not sure if acceptor_id and, accepted and deny_reason are really necessary. The first doesnt really matter, instead of accepted I would use local_user.activated (which i'm adding right now for email verification), and deny_reason can just be sent by email without storing. Well maybe that one makes sense for other admins to view.

**Comment by dessalines** [comment]
Still some work to do here, but the difficult part is done.

**Comment by Nutomic** [comment]
Trying to test it now. It looks like you also made it that login is required to view any content, didnt think that was part of this milestone. Only trying it on the instance for now, havent looked at the code yet.

- Its probably necessary to add an alternative frontpage which informs that the instance is private (instead of the small popup "this instance is private" which can easily be missed). Right now it just keeps loading forever.
- Next problem is that i cant login to my existing account on ds9, it gives error "email not verified". That shouldnt happen as it completely locks people out of their account. I set both email_verified and accepted_application true in the db, but still cant login (same error).
- On signup, it should probably show some message like "verify your email now so that you can login".
- The verification email is sent by your gmail address for some reason.
- The email verification link doesnt seem to work right, after clicking its again loading forever. It should show some confirmation like "your email has been verified", probably alongside the login form.

File: crates/api/src/local_user.rs
Comment (Nutomic) [inline crates/api/src/local_user.rs:978]: So clicking the verification link should directly log the user in, without password check? Thats not very secure, and besides it didnt seem to work for me.

File: crates/api/src/local_user.rs
Comment (Nutomic) [inline crates/api/src/local_user.rs:189]: This is also how i implemented it, but i dont think it makes sense. If there is an email being set, Lemmy should always send a verification mail, regardless of the `require_email_verification` setting. Cause its really useless to have an email with a typo in it. What `require_email_verification` should do is make the email field mandatory on signup (and prevent unsetting it in the user settings), nothing else.

File: crates/utils/src/settings/structs.rs
Comment (Nutomic) [inline crates/utils/src/settings/structs.rs:200]: Worth emphasizing that even with this setting, other instances can view all posts, comments etc from the instance via federation if it is enabled. And because of that, it  is easy to iterate through all posts and comments using `curl -H 'Accept: application/activity+json' https://ds9.lemmy.ml/post/x`. 

So maybe it should throw an error if both `private_instance` and` federation.enabled` are set (at least until #868 or #1576 are implemented).

**Comment by dessalines** [comment]
> Trying to test it now. It looks like you also made it that login is required to view any content, didnt think that was part of this milestone.

Only when `private instance` is set to true.

> Its probably necessary to add an alternative frontpage which informs that the instance is private (instead of the small popup "this instance is private" which can easily be missed). Right now it just keeps loading forever.

Hrm, it shouldn't do that, but I'll test again. Right now I have it so that it will redirect you to the login page if you get that error (just tested, yes this is working ). But I can add a "This instance is private" header to the login page.

> Next problem is that i cant login to my existing account on ds9, it gives error "email not verified". That shouldnt happen as it completely locks people out of their account. I set both email_verified and accepted_application true in the db, but still cant login (same error).

If a server suddenly turns on require_email_verification, I'm not sure the best way to not lock out existing users. That's a tough one. Maybe when that gets set to true, we need to update that `email_verified` column to true for every existing user ( even though that isn't actually the case ). I could use some help thinking of the best way to handle that.

> On signup, it should probably show some message like "verify your email now so that you can login".

It shows up popup that says "verification email sent", and that field is required when the site requires it, so we should be good there.

> The verification email is sent by your gmail address for some reason.

lemmy.ml is blocked by spamhaus so to even be able to test this, I had to use my gmail creds. Unfortunately I think we're gonna have to give up on postfix for lemmy.ml and just pay for an email service once this gets merged.

> The email verification link doesnt seem to work right, after clicking its again loading forever. It should show some confirmation like "your email has been verified", probably alongside the login form.

I tested that a few days ago, and it worked fine. It should only load for a second or two. I'll also make that isomorphic on the front end tho just in case. I'd be interested to see your console logs there when it failed.

File: crates/api/src/local_user.rs
Comment (dessalines) [reply crates/api/src/local_user.rs:978]: Hrm... yeah I spose that's correct. Dang, I just changed that respose to `LoginResponse` from an empty `VerifyEmailResponse`, but I can revert that. In that case, on a successful verify, I'll just redirect them to the login page.

File: crates/utils/src/settings/structs.rs
Comment (dessalines) [reply crates/utils/src/settings/structs.rs:200]: I can add that check to main.rs

**Comment by dessalines** [comment]
@Nutomic okay these fixes are deployed to ds9.lemmy.ml, you can test whenever. Note that you'll have to disable federation if you want to test private instances.

**Comment by Nutomic** [comment]
> Only when private instance is set to true.

Yes its no problem, i just didnt think we would implement it as part of this milestone.

> Hrm, it shouldn't do that, but I'll test again. Right now I have it so that it will redirect you to the login page if you get that error (just tested, yes this is working ). But I can add a "This instance is private" header to the login page.

Okay that works, but only when you initially open the site. When you are on the login page and navigate to another page (eg the main page, or communities page), it shows an error popup and loads forever. Also i noticed the mod log is still publicly visible for a private instance.

> If a server suddenly turns on require_email_verification, I'm not sure the best way to not lock out existing users. That's a tough one. Maybe when that gets set to true, we need to update that email_verified column to true for every existing user ( even though that isn't actually the case ). I could use some help thinking of the best way to handle that.

Not sure either. It could be enough to document this in the config, telling admins that they manually need to update this column for existing users.

> It shows up popup that says "verification email sent", and that field is required when the site requires it, so we should be good there.

Problem is that those popups are extremely easy to miss, as they are in a corner of the screen and disappear pretty quickly.

> lemmy.ml is blocked by spamhaus so to even be able to test this, I had to use my gmail creds. Unfortunately I think we're gonna have to give up on postfix for lemmy.ml and just pay for an email service once this gets merged.

Only if we enable email verification, which i doubt is gonna help much with spam. The registration applications are gonna be a lot more effective in that regard. But if there is an email service with a reasonable price, we can consider it.

> I tested that a few days ago, and it worked fine. It should only load for a second or two. I'll also make that isomorphic on the front end tho just in case. I'd be interested to see your console logs there when it failed.

Yes this time it worked fine. But when clicking the verification link for a second time (after already being verified), it again ends up loading forever (just with a small error popup). Maybe in cases like this, it should show the error(s) in place of the loading bar (and no popup).

**Comment by dessalines** [comment]
> Okay that works, but only when you initially open the site. When you are on the login page and navigate to another page (eg the main page, or communities page), it shows an error popup and loads forever. 

Not much I can do about that. The most important thing is the initial page load puts you in the right place.

> Also i noticed the mod log is still publicly visible for a private instance.

Looks like `GetModLog` doesn't even have an auth field. I'll add that.

> Not sure either. It could be enough to document this in the config, telling admins that they manually need to update this column for existing users.

I already added the previous solution to the code. I tested, and it works.

> Only if we enable email verification, which i doubt is gonna help much with spam

I agree, I'd rather not enable email verification. But it still would probably be worth it to get an email service, so that other notification emails and password resets can still get sent out without problems. Those are all currently broken now due to spamhaus.

> But when clicking the verification link for a second time (after already being verified), it again ends up loading forever (just with a small error popup). Maybe in cases like this, it should show the error(s) in place of the loading bar (and no popup).

I'll check this out on the front end.

**Review by Nutomic — APPROVED** [review]

**Pull request merged**
